### ISSUE TITLE
WeightedDL fails with Normalize

### ISSUE THREAD

**Issue by jimmiemunyi** [fastai/fastai#3795]
Please confirm you have the latest versions of fastai, fastcore, and nbdev before reporting a bug:
YES

fastai - 2.7.9
fastcore - 1.5.25
torch - 1.12.1
nbdev - 2.2.19

**Describe the issue**
When using `Datasets.weighted_dataloaders`, when I pass in `Normalize` in the `after_batch` section, I get an error. This doesn't happen when using `Datasets.weighted_dataloaders` without `Normalize` and also when I use `Datasets.dataloaders` with `Normalize` in the `after_batch`

This notebook describes the bug I am getting better and steps to reproduce
https://gist.github.com/jimmiemunyi/5a76b5c15ce40d43fb00e8768af0cf4b

**Expected behavior**
I can create a weighted dataloader and pass in the `Normalize` Transform. I am curious if I am using the `Normalize` wrongly

**Comment by TomPham97** [comment]
I tried these codes from the [Fastai data transformation documentation](https://docs.fast.ai/data.transforms.html#normalize) and it ran without raising any error:
```python
mean,std = [0.5]*3,[0.5]*3
mean,std = broadcast_vec(1, 4, mean, std)

dls = dsets.weighted_dataloaders(wgts=wgts, bs=bs, after_item=[Resize(img_size), ToTensor()],
                        after_batch=[IntToFloatTensor(), Normalize.from_stats(mean, std)])
```

I'm not entirely sure why this is the case, but you're welcome to look at [`Normalize` source code](https://github.com/fastai/fastai/blob/master/fastai/data/transforms.py#L364) and see if there's an issue.

**Comment by Ben-Karr** [comment]
Normalize tries to calculate `mean` and `std` if they are not given,. This happens when creating the loader, but the `wgts` argument is not set at that moment, so it can not pull a batch to do so and produces the error.
```python
@delegates()
class WeightedDL(TfmdDL):
    "Weighted dataloader where `wgts` is used for the training set only"
    def __init__(self, dataset=None, bs=None, wgts=None, **kwargs):
        super().__init__(dataset=dataset, bs=bs, **kwargs)
        wgts = array([1.]*len(dataset) if wgts is None else wgts)
        self.wgts = wgts/wgts.sum()
    …
    …
```
Swtiching the order gets rid of the error (for me):
```python
@delegates()
class WeightedDL(TfmdDL):
    "Weighted dataloader where `wgts` is used for the training set only"
    def __init__(self, dataset=None, bs=None, wgts=None, **kwargs):
        wgts = array([1.]*len(dataset) if wgts is None else wgts)
        self.wgts = wgts/wgts.sum()
        super().__init__(dataset=dataset, bs=bs, **kwargs)


    def get_idxs(self):
        if self.n==0: return []
        if not self.shuffle: return super().get_idxs()
        return list(np.random.choice(self.n, self.n, p=self.wgts))
```

**Comment by TomPham97** [comment]
Great! You should create a new Pull Request to fix this! ☺️

**Comment by TomPham97** [comment]
@Ben-Karr thanks for coming up with the solution. Would you be able to make a new Pull Request to implement this fix? ☺️
